### DISCUSSION TITLE
Formalizing the collaboration between agents

### DISCUSSION THREAD

**Discussion by fjuma** [wise-agents/wise-agents#179] [Ideas]
Just to follow up on our recent discussions, I think we should introduce the notion of a coordinator agent, a reasoning agent, and a final answer agent. I've described what this could look like below. Let me know what you think.

The coordinator agent will be responsible for coordinating the collaboration between agents in order to get to a final response. It can also handle adding/removing information from the global state that is shared by the agents.

Upon receiving a query, the coordinator agent will use an LLM along with the registry of agents to break the query down into sub-tasks that need to be achieved in order to determine the final response to the query. (Implementation wise, the coordinator could either consider the available agents as tools and determine the tool calls needed and execute them or this could also be implemented without tools.)

Upon receiving a response for a sub-task, the coordinator will ask the reasoning agent to determine whether or not the sub-task has actually been completed.

If the sub-task has been achieved, the next tool will be executed as usual.

If the sub-task has not been achieved, the coordinator could try to adjust the tool calls or rephrase the incomplete sub-task to dynamically re-adjust the planned tool calls. 

In addition to the reasoning agent being responsible for determining whether or not an agent has successfully completed its task, the coordinator will also use the reasoning agent to determine if we have reached a final response for the original query. If so, the coordinator will make use of the final answer agent to format the final response in the desired manner.

**Comment by maeste** [comment]
In general I agree with the design. The main focus in the design is considering the context needed by each agent (task, sub-task, coordinator, reasoning, and final answer producer) to pass this context (which maybe are tokens) between agents. Maybe we can/should consider also an algorithm/pattern to summarize and distill information to pass only relevant info for each step/call

**Reply by fjuma** [reply]
Right, that's a very good point. Will keep that in mind while working on the implementation.

**Comment by maeste** [comment]
> Just to follow up on our recent discussions, I think we should introduce the notion of a coordinator agent, a reasoning agent, and a final answer agent. I've described what this could look like below. Let me know what you think.

I was thinking more about this. Here and in all our previous discussions (and implementations also), we considered the coordinator agent as an LMM (or not) based agent that calls the other agents and waits for their response to move on to the next step. In your design, you went a bit further, introducing the concepts of reasoning and fina-answer-curator. Both of them are other two agents that the coordinator uses as yet another two tools providing some capabilities to solve part of the problem. 

This is definitely a possible implementation. I go further, saying it is definitely one of the implementations wise agents should implement and also most likely the first one they should provide.
That said I asked myself what a multi-agent system is providing with this implementation that makes a real difference compared to a complex LLM assistant or chatbot using a set of tools (APIs). Well, there is some obvious advantages we can provide:
* agents may (and it will be by design in our implementation) distributed. And with that I mean distributed in (bootable) pod which may be deployed in cloud
* Agents can use different LLM each. Which could be interesting for many scenarios having specialised LLM for some of the sub-goals
* Agents can operate in parallel (in theory at least, depends on the coordinator implementation)
* Wise(Agents) communicate asynchronously, favouring resilience and distribution
* The contxt/tokens sent to a specific tool may be not the full context of the conversation

But there are other aspects in which they are (with the coordinator implementation discussed in this thread) for which they are not so different from a monolithic LLM system
* Coordiantor keep all the  context
* Coordinator is single point of failure (of course can be clusterized)
* The collecting information/reasoning and also refining the steps needed is somehow sequential
* The last point means that the query and the coordinator's LLM knowledge also determines the action sequence more than by data. 

So, what are the other approaches I started thinking about?
The idea is to change the way we collect and use information and/or partial reasoning. Quick description of possible steps (to be discussed and refined):
1. The coordinator could determine all the sources of information/reasoning available (from the registry) that can somehow contribute to the solution. This determination can be based on agent self-description, which must be rich and possibly matched (vectorized) with the query.
2. The coordinator classifies the agents from the previous step into different groups such as context enrichment/data collection, context refinement/data analysis
3. All the agents in the first group are invoked in parallel. All of them enrich the context (in fact, the agents shared memory) with their pieces of information
4. Agents of the second group are launched in parallel (or in sequence where it makes sense). These agents use the shared memory as is (so with much more information than the single query) to get to a final answer
5. The coordinator (possibly using a tool/toolAgent for reasoning and curating the final answer) may decide to cycle on point 3 and 4

It's a very rough description, but it tries to mimic how humans approach a problem: search for pieces of information, group and organize them, and reason on them to get to an answer. This is much more challenging to implement in a monolith than in an agent-based architecture, so it could make a difference for us.

**Reply by fjuma** [reply]
> > Just to follow up on our recent discussions, I think we should introduce the notion of a coordinator agent, a reasoning agent, and a final answer agent. I've described what this could look like below. Let me know what you think.
> 
> I was thinking more about this. Here and in all our previous discussions (and implementations also), we considered the coordinator agent as an LMM (or not) based agent that calls the other agents and waits for their response to move on to the next step. In your design, you went a bit further, introducing the concepts of reasoning and fina-answer-curator. Both of them are other two agents that the coordinator uses as yet another two tools providing some capabilities to solve part of the problem.
> 
> This is definitely a possible implementation. I go further, saying it is definitely one of the implementations wise agents should implement and also most likely the first one they should provide. That said I asked myself what a multi-agent system is providing with this implementation that makes a real difference compared to a complex LLM assistant or chatbot using a set of tools (APIs). Well, there is some obvious advantages we can provide:
> 
> * agents may (and it will be by design in our implementation) distributed. And with that I mean distributed in (bootable) pod which may be deployed in cloud
> * Agents can use different LLM each. Which could be interesting for many scenarios having specialised LLM for some of the sub-goals
> * Agents can operate in parallel (in theory at least, depends on the coordinator implementation)
> * Wise(Agents) communicate asynchronously, favouring resilience and distribution
> * The contxt/tokens sent to a specific tool may be not the full context of the conversation

Another aspect that wasn't clear in my original description but have been experimenting with today is having the coordinator first use an LLM to break down the original query into sub-tasks using the description of the agents from the registry (so without using assistant tools) where each sub-task could be achieved by one or more agents.

The idea is then once these sub-tasks have been determined, then we can determine and execute the tools calls needed to achieve the sub-task. (Technically, the tools approach wouldn't be needed if we ask for the sub-tasks to be exectuable by a single agent. Then each sub-task would just be assigned to an agent.)

This initial query decomposition step helps to break down the problem into smaller pieces for which we then leverage the tools approach (or just assign a single agent to a sub-task).

This approach might have advantages over just using the tools directly to try to solve a complex query and makes it possible to analyze whether or not the sub-task was actually achieved.

That being said, I did realize that with this approach, sub-tasks would be executed sequentially when there's a potential for things to be done in parallel. Still thinking more about this part.

From the query decomposition I was experimenting with today, it seems like an LLM could break a query down in a way that mimics how humans approach the problem given a suitable description of agents (although the examples I tried were quite simple).

Here's an example query and agent descriptions that I experimented with and the sub-tasks that were determined by the coordinator:

```
Query: How do I prevent the following exception from occuring: Exception Details: java.lang.NullPointerException at com.example.ExampleApp.processData(ExampleApp.java:47)

Available Agents:
Agent Name: Agent4 Agent Description: This agent provides information about error messages using Source1
Agent Name: Agent5 Agent Description: This agent provides information about error messages using Source2
Agent Name: Agent6 Agent Description: This agent describes the underlying cause of a problem given information about the problem. This agent should be called after getting information about the problem using Agent4 and Agent5.
Agent Name: Agent7 Agent Description: This agent can be used to verify the information provided by other agents

Sub-Tasks determined by the LLM:
* Determine the meaning of the exception java.lang.NullPointerException using Agent4 and Agent5.
* Determine the underlying cause of the problem using Agent6 with information from Agent4 and Agent5.
* Verify the information about the underlying cause of the problem using Agent7.
* Determine how to prevent the exception from occurring based on the underlying cause of the problem using Agent6., 
``` 

> But there are other aspects in which they are (with the coordinator implementation discussed in this thread) for which they are not so different from a monolithic LLM system
> 
> * Coordiantor keep all the  context
> * Coordinator is single point of failure (of course can be clusterized)
> * The collecting information/reasoning and also refining the steps needed is somehow sequential
> * The last point means that the query and the coordinator's LLM knowledge also determines the action sequence more than by data.
> 
> So, what are the other approaches I started thinking about? The idea is to change the way we collect and use information and/or partial reasoning. Quick description of possible steps (to be discussed and refined):
> 
> 1. The coordinator could determine all the sources of information/reasoning available (from the registry) that can somehow contribute to the solution. This determination can be based on agent self-description, which must be rich and possibly matched (vectorized) with the query.
> 2. The coordinator classifies the agents from the previous step into different groups such as context enrichment/data collection, context refinement/data analysis
> 3. All the agents in the first group are invoked in parallel. All of them enrich the context (in fact, the agents shared memory) with their pieces of information
> 4. Agents of the second group are launched in parallel (or in sequence where it makes sense). These agents use the shared memory as is (so with much more information than the single query) to get to a final answer
> 5. The coordinator (possibly using a tool/toolAgent for reasoning and curating the final answer) may decide to cycle on point 3 and 4
> 
> It's a very rough description, but it tries to mimic how humans approach a problem: search for pieces of information, group and organize them, and reason on them to get to an answer. This is much more challenging to implement in a monolith than in an agent-based architecture, so it could make a difference for us.

I agree that it will be important to parallelize things where possible. 

From the query decomposition example above, it looks like an LLM could determine the steps proposed here given suitable agent descriptions but that being said, there might still be advantages to formalizing these steps explicitly. Will think about this some more. One thing I'm wondering is if the agent descriptions can be used to determine what steps can be run in parallel.

**Reply by fjuma** [reply]
> So, what are the other approaches I started thinking about? The idea is to change the way we collect and use information and/or partial reasoning. Quick description of possible steps (to be discussed and refined):
> 
> 1. The coordinator could determine all the sources of information/reasoning available (from the registry) that can somehow contribute to the solution. This determination can be based on agent self-description, which must be rich and possibly matched (vectorized) with the query.
> 2. The coordinator classifies the agents from the previous step into different groups such as context enrichment/data collection, context refinement/data analysis
> 3. All the agents in the first group are invoked in parallel. All of them enrich the context (in fact, the agents shared memory) with their pieces of information
> 4. Agents of the second group are launched in parallel (or in sequence where it makes sense). These agents use the shared memory as is (so with much more information than the single query) to get to a final answer
> 5. The coordinator (possibly using a tool/toolAgent for reasoning and curating the final answer) may decide to cycle on point 3 and 4
> 
> It's a very rough description, but it tries to mimic how humans approach a problem: search for pieces of information, group and organize them, and reason on them to get to an answer. This is much more challenging to implement in a monolith than in an agent-based architecture, so it could make a difference for us.


Just writing some thoughts here.

Upon receiving a request, the coordinator would determine groups of agents that should be executed and the order in which the groups should be executed.

When processing a response, once all the agents from the current group have completed, requests could be sent to the next group of agents or the coordinator could choose to cycle on a specific group.

**Reply by maeste** [reply]
This query and task split is great (in general what described in all your answer)

Let me clarify though the main point of my write up about an alternative approach. This is not much different from splitting goal and invoking agents perspective, but more on how single task answer is shared and used by other agents. With the current approach we are considering agents as request/response components and the caller of an agent wait for its response and eventually use it. The approach I'm proposing is removing the response (which become just an ack of task completed) and all the data acquired or generated by the agent would be put in a shared memory (the context). The main advantage of this approach is that all agents can use what other agents has already collected, moving from the coordinator to each single agent the know-how and the responsibility to decide what information present in the context ro use and how.

Does it make sense?

**Reply by fjuma** [reply]
The shared memory/context part makes sense. The part that's less clear to me though is if the coordinator is not waiting for a response from an agent/group of agents, how does an agent/group of agents know when to start its work. Are agents monitoring the shared context waiting for some specific thing to be available there before starting their work?

Just adding a note here since this last question was answered in our discussion today so adding the answer here too. The coordinator would still receive an ack of the task completed, the response just wouldn't contain all the info, that would be added to the shared context.

**Reply by maeste** [reply]
it's waiting for an ACK and it is responsible to move on calling another group of agents

**Comment by fjuma** [comment]
Just a quick update on this. I've submitted a PR for an initial coordinator implementation that executes agents in parallel in phases based on the discussions here:

https://github.com/wise-agents/wise-agents/pull/207
